Move General::buildPaginationElement >> DataSource

nitriques opened the pull request:
See commit details.

Fixes #2670

(Do not merge for now)

ChoboHub committed:
Remove the single output parameter (#2610)

This has been deprecated for some time now

Picked from 0c8dc7c
Picked from c524c3af4f

ChoboHub committed:
Remove associated count with section handle only

Keeping only the one with the section handle + field handle.
from :
        <entry id="6" greetings-test="0" greetings="0" />
        <entry id="5" greetings-test="0" greetings="0" />
        <entry id="4" greetings-test="0" greetings="0" />
to :
        <entry id="6" greetings-test="0" />
        <entry id="5" greetings-test="0" />
        <entry id="4" greetings-test="0" />

Picked from 7e4a829
Re #2610

Picked from b05cf17b27

nitriques committed:
Move General::buildPaginationElement >> DataSource

General is a already quite loaded. This method those not belongs there.

This commit replaces the implementation in General with a call to
DataSource. The General method is marked as deprecated. The General
method now calls the new implementation directly.

The new implementation uses an array to receive optional parameters via
an array, which makes the implementation a bit better.

Fixes #2670

nitriques committed:
Fix typos

As per review made by @michael-e on
c35d1a75d2675c466eb9f74804ba698b7af409a1

As per review made by @brendo on 1649f39fb49124387122d8abf6eab56af0abf72c

brendo commented on symphony/lib/toolkit/data-sources/class.datasource.section.php:
```diff
@@ -581,7 +581,9 @@ public function execute(array &$param_pool = null)
 
             if ($include_pagination_element) {
-                $pagination_element = General::buildPaginationElement();
+                $pagination_element = $this->buildPaginationElement(array(
```
Use `Datasource::` to access the method (because it's static).

nitriques replied:
Indeed.

brendo commented on symphony/lib/toolkit/class.general.php:
```diff
@@ -1494,6 +1494,7 @@ public static function createXMLDateObject($timestamp, $element = 'date', $date_
     /**
      * Construct an XML fragment that describes a pagination structure.
      *
+     * @deprecated @since Symohony 2.7.0. Use DataSource::buildPaginationElement()
```
Two typos, Symohony (Symphony) and DataSource (Datasource)

brendo commented on symphony/lib/toolkit/class.datasource.php:
```diff
@@ -241,6 +241,45 @@ public function emptyXMLSet(XMLElement $xml = null)
+     * @return XMLElement
+     *  the constructed XML fragment.
+     */
+    public static function buildPaginationElement(array $infos = array())
```
I feel the old signature was cleaner, what prompted the change?

nitriques replied:
In order to make function signature resilient to changes, its a good practice to try to replace parameter with defaults values with a "settings bags".

You can then

1. Specify the 3rd option only, and keep the default intact.
2. Add a new option without worrying about signatures/API compat
3. Extension could ship there could with new option and it should not break on old install!

It only needs to be documented

Those parameters was the perfect example of this code smell and I did wanted to try it out.

nitriques replied:
Most of the Widget would benefit from this... $value, $attributes.

brendo replied:
That is a really interesting perspective, thanks for sharing!

nitriques replied:
The same pattern exists in javascript too, where you have an "configuration object" that merges with defaults ;) Like most jQuery plugins

brendo commented on symphony/lib/toolkit/class.datasource.php:
```diff
@@ -241,6 +241,45 @@ public function emptyXMLSet(XMLElement $xml = null)
 
     /**
+     * Construct an XML fragment that describes a pagination structure.
+     * Do not call as a static function, since this is only temporary.
```
?

nitriques replied:
It SHOULD be static, but it can't right now (it was crashing on my end...) Maybe I need to revisit this.

michael-e commented on symphony/lib/toolkit/class.general.php:
```diff
@@ -1508,16 +1509,14 @@ public static function createXMLDateObject($timestamp, $element = 'date', $date_
      *  the constructed XML fragment.
      */
-    public static function buildPaginationElement($total_entries = 0, $total_pages = 0, $entries_per_page = 1, $current_page = 1)
+    public static function buildPaginationElement($total_entries = 0, $total_pages = 0, $entries_per_page = null, $current_page = 1)
```
Why `null`? Shouldn't it be `0` (because we assign integer values)?

nitriques committed:
Change the Server requirements

- Drop PHP5.3-5.5
- Drop MySQL 5.5

All of those version are not supported anymore

Picked from ae74bcd722
Picked from 3ea3cb53a9

nitriques committed:
Add grunt-cli as a dependency

This is needed to be able to run the build without having anything
installed globally.

The README got updated but we need to update the wiki:

https://github.com/symphonycms/symphony-2/wiki/Contributing-to-Symphony
https://github.com/symphonycms/symphony-2/wiki/Building-Symphony's-Assets
https://github.com/symphonycms/symphony-2/wiki/Building-Symphony

Re #2717

Picked from 499b82b76d
Picked from 893e29e818

nitriques committed:
Simplify the trailing slash rewrite condition

The second cond is not mandatory, since the first one will catch
everything.

Fixes #2730

Picked from 194a54696f
Picked from f308df9bf1

nitriques committed:
Add curl_error message (#2551)

* Add curl_error message

This caught my attention: curl_error actually contained the curl_errno

Picked from 6b87161c2f

* Replace curl_error by curl_errno

This fix was needed as per 6b87161

Picked from d7a92c0b1c

* Use strict comparison

Picked from 0d8fea67fa
Picked from d7084122fb
Picked from 00ddd22253

nitriques committed:
Add 'is (not) null' filter

We already provided the 'simple' SQL clause to check for nullable, but
it was not possible to combine null checks with another filter, like so:

earlier than today, is null

credits to @PascalPiche

Picked from 16a58f4e4b
Picked from e55d6db921

nitriques committed:
Get rid of selectize

We do not use it anymore

Picked from b366b1e6d9
Picked from 978990b9d6

jonmifsud committed:
Add delegates to allow custom sorting (#2758)

Sorting can not be manipulated by extensions.

Picked from 96b52644b6
Picked from 3001c91613

nitriques committed:
Update to jQuery 3.2.1

Also:

- Add jQuery migrate, to help with the migration
- Updated clndr to latest version
- Fix the code to remove as much warnings as possible
- Update uglify grunt task
- Add curl grunt task to download assets

Picked from 2457daf1ce
Picked from 6c7cc96a40

nitriques committed:
Add callbacks for filtering plugin

This allows extensions developers to customize the appearance of the
filtering results

Picked from 42fb1488b4
Picked from 72828dd023
Picked from 5285a5c20b
Picked from 500c101cd0

nitriques committed:
Make the default max log size to 10Mb (#2801)

This is to prevent over rotating the logs

Fixes #2757

Picked from 8a9a578
Picked from d69537b9e2

michael-e committed:
JSON to XML: Only add element attributes if useful (#2798)

Picked from 63a6fca
Picked from 8fbe6fb9aa

nitriques committed:
Revert to semi wildcard for uglify

Because the order of import is important

Picked from e86ca21f6a

nitriques committed:
Fix display (indendation) of stack traces

Picked from b12dc1059d

nitriques committed:
Remove undefined variable

Picked from 16dbcc243f

nitriques committed:
Fix jshint error

Picked from c7a13b05ce

nitriques committed:
Remove bottom margin if last-child

Picked from 64402fdf79

nitriques committed:
Add a default security.txt file (#2808)

This file allows security researchers to find us quickly if they find a
security related problem.

See https://securitytxt.org

Fixes #2807

Picked from 7cad4b71e1

nitriques committed:
Make DS params in XML optional (#2805)

This commits add a new parameter in data sources to control if the
output parameters should be added in the xml output.

By default, data sources that do not have the dsParamPARAMXML property
set will include their output parameters in the XML, since this is their
current behavior. If the dsParamPARAMXML exists on the object, it only
needs to be different from 'yes' to be excluded from the XML.

Fixes #784

Picked from d9ea14464b

nitriques committed:
Better error handling and disabled display of exception data (#2597)

* Fix doc: E_NOTICE are loggued c8ff3dc

The doc stated that E_NOTICE were ignored while the code only ignores
E_STRICT (see line 423)

* Make renderHtml protected ebc5ac7

This method should never get called directly.
Nothing ever calls it in the current base code.

* Disable GenericExceptionHandler before auth d53b221

Those changes makes sure we do not leak exception information before
doing the authentication. It also makes sure that exception raised in
the exception handler won't leak info if the GenericExceptionHandler is
disabled.

The same logic is also applied in the shutdown function.

Instead of masking the error like we did, the logic is applied only in
the render part of things, which should make it easier for us to debug.
It also removes the wrong HTTP status that we were returning (500
instead of 404)

Picked from dbcf209

* Add GenericExceptionHandler::$logDisabled dec526d

This implementation copies what was already present in the
GenericErrorHandler, i.e. skip logging if disabled = true.

This also prevents errors caught in GenericErrorHandler to be also
loggued a second time in the GenericExceptionHandler::handler method.

* Check GenericExceptionHandler::$enable in subclass 20f37aa

Both SymphonyErrorPageHandler and DatabaseExceptionHandler needed care:

SymphonyErrorPageHandler:
- throwCustomError must not touch GenericExceptionHandler::$enabled
  since it allows data leakage
- defaults the diabled and not template cases to
  GenericExceptionHandler::render

DatabaseExceptionHandler:
- Filter out data if the Handler is not enabled. This handler was
  already getting the right template since it uses the same as the
parent.

GenericExceptionHandler:
- refactored the headers part to share code in SymphonyErrorPageHandler

* Made handlers final 8d84a7d

Both the GenericErrorHandler and the GenericExceptionHandler's handler
function should not be overriden.

* FrontendPageNotFoundExceptionHandler bypass 809f927

FrontendPageNotFoundExceptionHandler can bypass the new anti-leak
system since it's a legitimate message.

Picked from 3da6eeb9a8

michael-e committed:
Data Sources: Allow to output system IDs only (#2806)

As discussed in #784, it may be useful to output system IDs only.

Picked from f843d96
Picked from 5ccc9f04b4

nitriques committed:
Add $changing_email/password to AuthorPreEdit

This allows developers to act on when those values changes

Picked from 775fbae182

nitriques committed:
Add $changing_email/password to AuthorPostEdit

This allows developers to act only when those values changes¸
and that the author was properly saved.

Re 775fbae182
https://github.com/symphonycms/symphony-2/commit/775fbae18299f72a5e6ff6b0f271c9e3b804cdeb#r28055757

Picked from 1954393d1a

nitriques committed:
Add the Content-ID header on attachments (#2617)

* Add the Content-ID header on attachments d7abb0ecceba646eb235cdaa6a703b4cdb639ce3

This allows developer to reference attachments in the html body to allow
images to be viewed offline.

You can use it as follow:

```html
<img src="cid:name-of-the-file.ext" />
```

Where the name-of-the-file.ext part matches the actual file name of the
file. Notice the `cid:` prefix, which makes it work!

Picked from 2b75aa7b70

* Add a flag to enable Content-ID header field 5483f71e075df0458798c2b0c1e2b13bb499fb37

* Let the developer controls the disposition value c3a6a270771508461019aa5a3d760029c39dd054

* Let the developer controls the content-id value 87103096d5b06316db94d03f82eaa8c7aa0714d6

* Remove null keys from array and add isset() 24786e17c82426b14ca50e428838f1054a6a3974

Also remove the old undocumented format

* Add missing commas 5c759921ae4b8b74f88672542f7436b0e32cf9e0

Picked from 895d7cb494

nitriques committed:
Add missing Resources delegates (#2813)

Both events and data sources were missing the PostDelete delegates.

No delegates were fired when the resources were deleted from the index
page.

The handle of the resource is added to PreDelete also.

Re https://github.com/symphonists/globalresourceloader/issues/13

Picked from 8104058
Picked from 2622f49061

nitriques committed:
Remove parameters as arguments (#2804)

* Remove parameters as arguments d3c31c39

i.e. using $param in XSLT without declaring it.

Fixes #2338

* Redistribute responsibilities between Page/Process 0614ff53

We used to store the xml and xsl values both in the XSLTPage AND in the
XSLTProcess. This is redundant and totally not needed.
This change makes only the Page responsible to hold the xml and xslt
data, which are simply passed when the XSLTProcess needs to do it job.

The only reason why the Process would need to hold the data is when
there is an error. For this purpose, the "context" is managed manually
instead of exposing it publicly.

On the other hand, the Page was holding data belonging to the Process:
params and functions. Since the page does not use the parameters anymore
and it does not register any functions, this data has been moved to the
process. Upon instanciation, the Page gets a process, so developers can
just forward their Page calss to Page->Proc.

The end goal was to not expose any XSLT parameters or function in the
XSLTPage, which is now achieved.

Picked from f715cf2638

nitriques committed:
Remove the vendor folder

This folder should be created by the developer or via the packager

Picked from e332b8458a
Picked from 8683955981

nitriques committed:
Remove static scope
